### PR TITLE
wpewebkit,webkitgtk: Bump to version 2.46.1

### PR DESCRIPTION
**Pull request by psaavedra**
* Create new recipe for version 2.46.X
* Delete recipes for version 2.42.5
    
Changes:
    
* Skia is used instead of Cairo for 2D rendering, and GPU rendering is enabled by default.
* Offscreen Canvas is now enabled by default.
* Add support for system tracing using Sysprof.
* Add new API to load settings from a config file.
* Add a new setting to toggle 2D Canvas acceleration (enabled by default).
* Undeprecate console messages API and make it available in the 2.0 API.
    
Release Notes:

* https://wpewebkit.org/release/wpewebkit-2.46.0.html
* https://wpewebkit.org/release/wpewebkit-2.46.1.html
    
Also CI updates in the build tests jobs definitions:
    
* Update test jobs definitions to primary build 2.46.X
* Remove preset for 2.42 and add a new one for 2.42

### CODE REVIEW

**Review by aperezdc — COMMENTED** [review]
@psaavedra Changes LGTM in general. We may want to add this somewhere in `wpewebkit.inc`, though:

```
EXTRA_OECMAKE += "-DENABLE_WPE_PLATFORM=OFF"
```

(See [this message in the mailing list](https://lists.webkit.org/pipermail/webkit-wpe/2024-September/000674.html), I forgot to make that option disabled by default in the release branch.)

**Comment by aperezdc** [comment]
> @psaavedra Changes LGTM in general. We may want to add this somewhere in `wpewebkit.inc`, though:
> 
> ```
> EXTRA_OECMAKE += "-DENABLE_WPE_PLATFORM=OFF"
> ```
> 
> (See [this message in the mailing list](https://lists.webkit.org/pipermail/webkit-wpe/2024-September/000674.html), I forgot to make that option disabled by default in the release branch.)

Actually, that won't work, there's an unconditional `SET_AND_EXPOSE_TO_BUILD()`, I'm making a patch.

**Comment by aperezdc** [comment]
> > @psaavedra Changes LGTM in general. We may want to add this somewhere in `wpewebkit.inc`, though:
> > ```
> > EXTRA_OECMAKE += "-DENABLE_WPE_PLATFORM=OFF"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > (See [this message in the mailing list](https://lists.webkit.org/pipermail/webkit-wpe/2024-September/000674.html), I forgot to make that option disabled by default in the release branch.)
> 
> Actually, that won't work, there's an unconditional `SET_AND_EXPOSE_TO_BUILD()`, I'm making a patch.

Here's the patch needed to be able to toggle the option: https://github.com/WebKit/WebKit/pull/33819

**Comment by psaavedra** [comment]
In the GTK build: | --   No package 'sysprof-capture-4' found. Is sysprof really a mandatory depends?  Can we set this depends off?

**Comment by aperezdc** [comment]
> @aperezdc , I have a doubt for the GTK build: | -- No package 'sysprof-capture-4' found. Is sysprof really a mandatory depends? Can we set this depends off?

You can go with `-DUSE_SYSTEM_SYSPROF_CAPTURE=OFF` to use the bundled copy of the library, or `-DUSE_SYSPROF_CAPTURE=OFF` to completely disable the feature.

**Comment by psaavedra** [comment]
we are getting a build error for the WebKitGTK build:

```
2024-09-19T10:40:08.8312493Z | /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/EligibilityResult.h:35:5: error: expected identifier before numeric constant
2024-09-19T10:40:08.8314249Z |    35 |     Success, 
2024-09-19T10:40:08.8314565Z |       |     ^~~~~~~
2024-09-19T10:40:08.8316408Z | /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/EligibilityResult.h:35:5: error: expected '}' before numeric constant
2024-09-19T10:40:08.8319541Z | In file included from /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/EligibilityResultInlines.h:28,
2024-09-19T10:40:08.8322401Z |                  from /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/IsoHeapInlines.h:31,
2024-09-19T10:40:08.8325041Z |                  from /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/WTF/Headers/wtf/TZoneMallocInlines.h:58,
2024-09-19T10:40:08.8327893Z |                  from /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/webkitgtk-2.46.0/Source/WebKit/UIProcess/gtk/PointerLockManagerX11.cpp:34:
2024-09-19T10:40:08.8330867Z | /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/EligibilityResult.h:34:28: note: to match this '{'
2024-09-19T10:40:08.8332470Z |    34 | enum class EligibilityKind {
2024-09-19T10:40:08.8332889Z |       |                            ^
2024-09-19T10:40:08.8334978Z | /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/EligibilityResult.h:35:5: error: expected unqualified-id before numeric constant
2024-09-19T10:40:08.8336719Z |    35 |     Success,
2024-09-19T10:40:08.8337030Z |       |     ^~~~~~~
2024-09-19T10:40:08.8339221Z | /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/EligibilityResult.h:44:22: warning: unnecessary parentheses in declaration of 'EligibilityKind' [-Wparentheses]
2024-09-19T10:40:08.8341210Z |    44 |     EligibilityResult(EligibilityKind);
2024-09-19T10:40:08.8341683Z |       |                      ^~~~~~~~~~~~~~~~~
2024-09-19T10:40:08.8343497Z | /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/EligibilityResult.h:44:22: note: remove parentheses
2024-09-19T10:40:08.8345154Z |    44 |     EligibilityResult(EligibilityKind);
2024-09-19T10:40:08.8345627Z |       |                      ^~~~~~~~~~~~~~~~~
2024-09-19T10:40:08.8346133Z |       |                      -               -
2024-09-19T10:40:08.8348328Z | /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/EligibilityResult.h:44:23: error: field 'EligibilityKind' has incomplete type 'EligibilityResult<Config>'
2024-09-19T10:40:08.8350293Z |    44 |     EligibilityResult(EligibilityKind);
2024-09-19T10:40:08.8350762Z |       |                       ^~~~~~~~~~~~~~~
2024-09-19T10:40:08.8353022Z | /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/EligibilityResult.h:41:8: note: definition of 'struct EligibilityResult<Config>' is not complete until the closing brace
2024-09-19T10:40:08.8355020Z |    41 | struct EligibilityResult {
2024-09-19T10:40:08.8355423Z |       |        ^~~~~~~~~~~~~~~~~
2024-09-19T10:40:08.8357270Z | /home/bot/yocto-webkit-scarthgap/builds/raspberrypi3-mesa-webkitgtk/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.46.0/build/bmalloc/Headers/bmalloc/EligibilityResult.h:45:30: error: expected ')' before '<' token
2024-09-19T10:40:08.8358997Z |    45 |     EligibilityResult(IsoPage<Config>*);
2024-09-19T10:40:08.8359461Z |       |                      ~       ^
2024-09-19T10:40:08.8359896Z |       |                              )
```

I am not familiar with this code. The error indicates that `Success` is being treated as a numeric constant or macro, possibly because Success has been defined somewhere else, potentially as a macro or constant.

**Comment by psaavedra** [comment]
Likely the issue could be related with this flag:

```
//Whether to enable support for the X11 windowing target.
ENABLE_X11_TARGET:BOOL=ON
```

**Comment by aperezdc** [comment]
Maybe you will want to pick [WebKitGTK 2.46.1](https://webkitgtk.org/2024/09/30/webkitgtk2.46.1-released.html) and wait for the corresponding WPE release 😉.

File: conf/templates/template/presets/wpe-2_46.conf [hunk 1]
```diff
@@ -1,7 +1,7 @@
 PREFERRED_PROVIDER_virtual/wpebackend = "wpebackend-fdo"
 PREFERRED_PROVIDER_virtual/libwpe = "libwpe"
 PREFERRED_VERSION_cog = "0.18.%"
-PREFERRED_VERSION_libwpe = "1.14.%"
+PREFERRED_VERSION_libwpe = "1.16.%"
 PREFERRED_VERSION_wpebackend-fdo = "1.14.%"
 PREFERRED_VERSION_wpebackend-rdk = "1.20200213"
-PREFERRED_VERSION_wpewebkit = "2.42.%"
+PREFERRED_VERSION_wpewebkit = "2.44.%"
```
Comment (aperezdc) [inline conf/templates/template/presets/wpe-2_46.conf:7]: Typo:

```suggestion
PREFERRED_VERSION_wpewebkit = "2.46.%"
```

File: recipes-browser/wpewebkit/wpewebkit.inc [hunk 2]
```diff
@@ -6,10 +6,12 @@ BUGTRACKER = "https://bugs.webkit.org/"
...
 LIC_FILES_CHKSUM = "file://Source/WebCore/LICENSE-LGPL-2.1;md5=a778a33ef338abbaf8b8a7c36b6eec80 "
 
+REQUIRED_DISTRO_FEATURES = "opengl"
+
 DEPENDS:append = " \
     libsoup \
-    bison-native gperf-native harfbuzz-native libxml2-native ccache-native ninja-native ruby-native cairo \
-    fontconfig freetype glib-2.0 harfbuzz icu jpeg pcre sqlite3 zlib libpng \
+    bison-native gperf-native harfbuzz-native libxml2-native ccache-native ninja-native ruby-native \
+    fontconfig freetype glib-2.0 harfbuzz icu jpeg pcre sqlite3 zlib libinput libpng libtasn1 \
```
Comment (aperezdc) [inline recipes-browser/wpewebkit/wpewebkit.inc:14]: The `libinput` package should be only needed with `-DENABLE_WPE_PLATFORM=ON`, and the build recipe does not use that.

File: recipes-browser/wpewebkit/wpewebkit_2.46.0.bb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+require wpewebkit.inc
+require conf/include/devupstream.inc
+
+FILESEXTRAPATHS:prepend := "${THISDIR}/${PN}:"
+
+SRC_URI = "https://wpewebkit.org/releases/${BPN}-${PV}.tar.xz;name=tarball \
+           file://0001-CMake-WPE-ENABLE_WPE_PLATFORM-cannot-be-toggled.patch \
```
Comment (aperezdc) [inline recipes-browser/wpewebkit/wpewebkit_2.46.0.bb:7]: You can use [WPE WebKit 2.46.1](https://wpewebkit.org/release/wpewebkit-2.46.1.html) now, and drop this patch that has been included in the release.

**Review by aperezdc — CHANGES_REQUESTED** [review]

**Comment by psaavedra** [comment]
> Likely the issue could be related with this flag:
> 
> ```
> //Whether to enable support for the X11 windowing target.
> ENABLE_X11_TARGET:BOOL=ON
> ```

issue: https://bugs.webkit.org/show_bug.cgi?id=278858

File: recipes-browser/wpewebkit/wpewebkit_2.46.1.bb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+FILESEXTRAPATHS:prepend := "${THISDIR}/${PN}:"
+
+SRC_URI = "https://wpewebkit.org/releases/${BPN}-${PV}.tar.xz;name=tarball"
+
+SRC_URI[tarball.sha256sum] = "1e0aaf870f36001c42b1ce5a2027b4101bed878746e437cc6d6fed0693afe9ad"
+
+SRC_URI:class-devupstream = "git://github.com/WebKit/WebKit.git;protocol=https;branch=main"
+
+SRCREV:class-devupstream = "8db1e051cf4a6d7877dd51b81b478e8367d3f052"
```
Comment (Qingwu-Li) [inline recipes-browser/wpewebkit/wpewebkit_2.46.1.bb:12]: eacf9977760d4ea5dfcb736be73ae48c7df16f27  ？

Comment (psaavedra) [reply recipes-browser/wpewebkit/wpewebkit_2.46.1.bb:12]: that tags belongs to the wpewebkit-2.46.1 branch. Notice that the SRC_URI:class-devupstream points to the `main` branch what is the default upstream branch for development. In general, I use the latest common commit between the release branch and main (`git merge-base main wpewebkit-2.46.1`) to set this value.

File: recipes-browser/webkitgtk/webkitgtk/fix-bmalloc-armhf.patch [hunk 5]
```diff
@@ -0,0 +1,19 @@
+From: Alberto Garcia <berto@igalia.com>
```
Comment (aperezdc) [inline recipes-browser/webkitgtk/webkitgtk/fix-bmalloc-armhf.patch:1]: @bertogg Any chance you could send this patch upstream?

**Review by aperezdc — CHANGES_REQUESTED** [review]
I still have a couple of comments.

Also, the build is currently broken with `-DENABLE_VIDEO=OFF` (both in `main` and in the 2.46 release branch); there are patches being worked on for this.

File: recipes-browser/webkitgtk/webkitgtk_2.46.1.bb [hunk 6]
```diff
@@ -96,6 +97,9 @@ EXTRA_OECMAKE = " \
                  -DENABLE_INTROSPECTION=OFF \
                  -DENABLE_GTKDOC=OFF \
                  -DENABLE_MINIBROWSER=ON \
+                 -DUSE_SYSPROF_CAPTURE=OFF \
+                 -DENABLE_SAMPLING_PROFILER=OFF \
+                 -DENABLE_WEBASSEMBLY=ON \
```
[the remark this answers is not in the document]
Comment (psaavedra) [reply recipes-browser/webkitgtk/webkitgtk_2.46.1.bb:102]:
```
                 -DENABLE_SAMPLING_PROFILER=OFF \
￼                 -DENABLE_WEBASSEMBLY=ON \
```
This is a mistake. They come from some back and forth doing tests. They should be not here. 

Thanks for 2x👁️

File: recipes-browser/webkitgtk/webkitgtk_2.46.1.bb [hunk 7]
```diff
@@ -81,6 +81,7 @@ PACKAGECONFIG[libhyphen] = "-DUSE_LIBHYPHEN=ON,-DUSE_LIBHYPHEN=OFF,libhyphen"
 PACKAGECONFIG[libsecret] = "-DUSE_LIBSECRET=ON,-DUSE_LIBSECRET=OFF,libsecret"
 PACKAGECONFIG[opengl] = "-DUSE_OPENGL_OR_ES=ON,-DUSE_OPENGL_OR_ES=OFF,virtual/libgl"
 PACKAGECONFIG[openjpeg] = "-DUSE_OPENJPEG=ON,-DUSE_OPENJPEG=OFF,openjpeg"
+PACKAGECONFIG[skia] = "-DUSE_SKIA=ON,-DUSE_SKIA=OFF,"
```
[the remark this answers is not in the document]
Comment (psaavedra) [reply recipes-browser/webkitgtk/webkitgtk_2.46.1.bb]: OK. Let's forget about this flag and rely on the CMake rules.

File: recipes-browser/wpewebkit/wpewebkit_2.46.1.bb (hunk 4 above)
Comment (psaavedra) [reply recipes-browser/wpewebkit/wpewebkit_2.46.1.bb]: @Qingwu-Li @aperezdc Ok, your argument has changed my point of view. I have added this additional commit: https://github.com/Igalia/meta-webkit/pull/509/commits/abd8ebd4425ed48c0b6b91a700c3a7f1e01d6e42, as we were using devupstream differently for referencing the upstream top-of-tree

**Comment by psaavedra** [comment]
> I still have a couple of comments.
> 
> Also, the build is currently broken with `-DENABLE_VIDEO=OFF` (both in `main` and in the 2.46 release branch); there are patches being worked on for this.

Could you be more specific? I see https://github.com/WebKit/WebKit/pull/35004  which is being about to be landed which this partially depends of 2f4378eb7afab. Is this what you are making a reference in you previous comment?

**Review by aperezdc — APPROVED** [review]
Thanks for all the updates to the PR, LGTM!

**Comment by aperezdc** [comment]
> > I still have a couple of comments.
> > Also, the build is currently broken with `-DENABLE_VIDEO=OFF` (both in `main` and in the 2.46 release branch); there are patches being worked on for this.
> 
> Could you be more specific? I see [WebKit/WebKit#35004](https://github.com/WebKit/WebKit/pull/35004) which is being about to be landed which this partially depends of 2f4378eb7afab. Is this what you are making a reference in you previous comment?

All these three patches to be able to build with multimedia support disabled:

- https://github.com/WebKit/WebKit/pull/34923
- https://github.com/WebKit/WebKit/pull/35004 (this is the one you linked above)
- https://github.com/WebKit/WebKit/pull/35147

I have backported them all into the `webkitglib/2.46` branch:

- https://github.com/WebKit/WebKit/commit/6b59562904cd85cebdd91106308d35c5e112858d
- https://github.com/WebKit/WebKit/commit/99989ffdf40b0290d9129a76a5128987f7299728
- https://github.com/WebKit/WebKit/commit/3f520c14f646445b2b242ace42a8aadd0d834a3b

If you are going to import the patches into `meta-webkit`, pick them from the release branch, because the second one has some additional fixes by me.

**Comment by blino** [comment]
> If you are going to import the patches into `meta-webkit`, pick them from the release branch, because the second one has some additional fixes by me.

About the `ENABLE(MEDIA_SESSION)`, I indeed left this part out in commit 2f4378eb7afab.
@aperezdc : are you going to push this to main branch too?

**Comment by aperezdc** [comment]
@psaavedra And one more fix for ARM NEON (needed e.g. for the Raspberry Pi 4) → https://github.com/WebKit/WebKit/pull/35244

@blino I will push the `ENABLE(MEDIA_SESSION)` bits to `main`, yes.

**Comment by psaavedra** [comment]
Thanks @aperezdc  I will take care of those patches and I will add it in a subsequent PR. 🙇🏼

**Pull request merged**